### PR TITLE
Send-to-device messages over federation

### PR DESCRIPTION
**Pull request by neilalexander**
This PR updates the federation sender to send send-to-device messages over federation. (That's a tongue-twister!)

### CODE REVIEW

File: federationsender/consumers/eduserver.go [hunk 1]
```diff
@@ -26,50 +29,124 @@ import (
...
+// send-to-device events topic from the EDU server.
+func (t *OutputEDUConsumer) onSendToDeviceEvent(msg *sarama.ConsumerMessage) error {
+	// Extract the send-to-device event from msg.
+	var ote api.OutputSendToDeviceEvent
+	if err := json.Unmarshal(msg.Value, &ote); err != nil {
+		log.WithError(err).Errorf("eduserver output log: message parse failed (expected send-to-device)")
+		return nil
+	}
+
+	// only send typing events which originated from us
```
Comment (Kegsay) [inline federationsender/consumers/eduserver.go:94]: Typo :)

File: federationsender/consumers/eduserver.go [hunk 2]
```diff
@@ -26,50 +29,124 @@ import (
...
+	if err != nil {
+		log.WithError(err).WithField("user_id", ote.UserID).Error("Failed to extract domain from send-to-device destination")
+		return nil
+	}
+
+	// Generate a random message ID for idempotency
+	chars := "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"
+	random := rand.New(rand.NewSource(time.Now().UnixNano()))
+	messageID := make([]byte, 32)
+	for i := range messageID {
```
Comment (Kegsay) [inline federationsender/consumers/eduserver.go:115]: Just use https://godoc.org/github.com/matrix-org/util#RandomString

Comment (neilalexander) [reply federationsender/consumers/eduserver.go:115]: Didn't realise we had that - thanks!

**Comment by neilalexander** [comment]
Failing test is `Outbound federation requests missing prev_events and then asks for /state_ids and resolves the state`.

**Pull request merged**